cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -803,6 +805,100 @@ public void suspendResumeSchedule(String namespaceId, String appId, String sched
   }
 
   /**
+   * Add a schedule to an application.
+   *
+   * @param applicationId the application id for which the schedule needs to be added
+   * @param scheduleSpec the schedule specification
+   * @throws NotFoundException when application is not found
+   * @throws SchedulerException on an exception when updating the schedule
+   */
+  public void addSchedule(ApplicationId applicationId, ScheduleSpecification scheduleSpec)
+    throws NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationSpecification appSpec = store.getApplication(applicationId);
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(applicationId);
+    }
+
+    String scheduleName = scheduleSpec.getSchedule().getName();
+    ScheduleSpecification existingScheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (existingScheduleSpec != null) {
+      throw new AlreadyExistsException(
+        new ScheduleId(applicationId.getNamespace(), applicationId.getApplication(), scheduleName));
+    }
+
+    // TODO: Figure out if there a way to make the scheduler update and store update transactional
+    ProgramType programType = ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
+    String programName = scheduleSpec.getProgram().getProgramName();
+    ProgramId programId = new ProgramId(applicationId, programType, programName);
+    scheduler.schedule(programId, scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule(),
```

Comment:
so the program type must be passed in twice? Please file an improvement Jira to consolidate that in the future and add a TODO